### ISSUE TITLE
Change types with ored compare_keys?

### ISSUE THREAD

**Issue by vchav73** [Yelp/elastalert#459]
I have a case where I need to alert on changes across any of multiple keys or if a number field exceeds a threshold. I'm still new to elastalert and it's not clear to me how to do this. Can you put multiple rule types in a single rule file or do I need to generate a separate rule file for each key I need to check? Would it be best to create my own rule type?

**Comment by Qmando** [comment]
One rule type per file. For the number field:
```
type: any
filter:
- range:
    field:
      gt: 10 # Greater than
```

for the change rule, you can unfortunately only monitor one field per rule. 

If your data is constantly coming in, there is something you can do that is similar to what you want, but not quite. You can set up a multi-field query key, and paired with a flatline type, will alert you when any combination of fields disappears. For example,
```
type: flatline
threshold: 1
query_key: ["username", "ip_address", "status"]
timeframe:
  days: 1
```
would alert you if any specific combination of (username, ip, status) appeared and then did not show up again for 24 hours.

The new_term rule exists for alerting on the opposite, a first ever occurrence of a new value. But you can only do that for independent fields, not regarding compound fields.

**Comment by vchav73** [comment]
Thanks for the explanation and examples.

Is creating a custom rule type also an option? I'm wondering partly for efficiency. Would a custom rule generate less elasticsearch queries than multiple rules files?

**Comment by Qmando** [comment]
Absolutely! http://elastalert.readthedocs.org/en/latest/recipes/adding_rules.html

Basically, you implement a class that stays persistent while ElastAlert runs, implements a function ``add_data`` (or ``add_terms_data`` and ``add_count_data``, not in the guide, look at other rule types) which takes a list of dictionary events, and then there is a function to ``add_match``.

You can of course subclass the existing rule types, which have some useful features like ``EventWindow`` which will automatically sort and store events based on a timeframe.
